keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,823 @@
+So, we limit the size of `succeededIndexes` to 64Ki.
+
+## Design Details
+
+### Job API
+
+We extend the Job API to set different policies by which a Job can be declared as succeeded.
+
+```golang
+type JobSpec struct {
+	...
+	// SuccessPolicy specifies the policy when the job can be declared as succeeded.
+	// If empty, the default behavior applies - the job is declared as succeeded
+	// only when the number of succeeded pods equals to the completions.
+	// Specified it must be immutable and works only for the Indexed jobs.
+	//
+	// This field is alpha-level. To use this field, you must enable the
+	// `JobSuccessPolicy` feature gate (disable by default).
+	// +optional
+	// +listType=atomic
+	SuccessPolicy *SuccessPolicy
+}
+
+// SuccessPolicy describes when a Job can be declared as succeeded based on the success of some indexes.
+type SuccessPolicy struct {
+	// Criteria represents the list of alternative criteria for declaring the jobs 
+	// as successful before its completion. Once any of the criteria are met,
+	// the "SuccessCriteriaMet" condition is added, and the lingering pods are removed.
+	// The terminal state for such a job has the "Complete" condition.
+	// Additionally, these criteria are evaluated in order; Once the job meets one of the criteria,
+	// other criteria are ignored.
+	//
+	// +optional
+	Criteria []SuccessPolicyCriterion
+}
+
+// SuccessPolicyCriteria describes a criterion for declaring a Job as succeeded.
+// Each criterion must have at least one of the "succeededIndexes" or "succeededCount" specified.
+type SuccessPolicyCriterion struct {
+	// SuccessPolicyCriterion specifies the set of indexes 
+	// which need to be contained in the actual set of succeeded indexes for the job.
+	// The list of indexes must come within 0 to ".spec.completions" and
+	// must not contain duplicates. At least one element is required. 
+	// The indexes are represented as decimal integers
+	// separated by commas. The numbers are listed in increasing order. Three or
+	// more consecutive numbers are compressed and represented by the first and
+	// last element of the series, separated by a hyphen.
+	// For example, if the completed indexes are 1, 3, 4, 5 and 7, they are
+	// represented as "1,3-5,7".
+	// 
+	// +optional
+	SucceededIndexes *string
+
+	// SucceededCount specifies the minimal size of the actual set of the succeeded indexes
+	// for the job.
+	// When specified it needs to be positive integer.
+	//
+	// +optional
+	SucceededCount *int
+	
+	// SuccessCriterionMetReasonSuffix specifies the CamelCase suffix of the reason for the "SuccessCriteriaMet" condition.
+	// Once the job meets this criterion, "JobSuccessPolicy" and the specified suffix is combined with "As".
+	// For example, if specified suffix is "LeaderSucceeded", it is represented as "JobSuccessPolicyAsLeaderSucceeded".
+	//
+	// +optional
+	SuccessCriterionMetReasonSuffix *string
```

Comment:
Can it be the whole reason, as opposed to a suffix?

Reply:
`SetConditionReason` would address the first concern.

A suffix just makes it harder to process from another controller (or another API). 

See for example how JobSet plans to use the reason https://github.com/kubernetes-sigs/jobset/pull/381/files#diff-95ff3f969c4dd0a07a5d8082df42bcb6f8f985e07bbe517b306635e397b4fca2R99

The user would have to remember to set `XYZ` in the Job spec, but `PodFailurePolicyXYZ` in the Jobset spec.

Reply:
Uhm, I agree with decreasing the value to add this to reason if we add a prefix since the controller needs to divide reasons by a separator (here is "As").

I agree that to make it more machine-readable, `SetConditionReason` would be better.

Reply:
I updated this API with `SetSuccessCriteriaMetReason`.